File: CMPS101/asg2/Lex.c
/**
   Name: Sean McGrath
   ID: ssmcgrat
   Assignment: CMPS 101 Programming Assingment 2

   Lex.c
*/

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "List.h"

#define MAX_LEN 160

int main(int argc, char* argv[]){

   int count=0;
   FILE *in, *out;
   char line[MAX_LEN];
   char array[MAX_LEN][80];
   char* token = NULL;

   for(int i=0; i<MAX_LEN; i++){
      strcpy(array[i], "");
   }

   List L = newList();
  
  // check command line for correct number of arguments
   if(argc != 3){
      printf("Usage: %s <input file> <output file>\n", argv[0]);
      exit(1);
   }

   // open files for reading and writing
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");
   if(in==NULL){
      printf("Unable to open file %s for reading\n", argv[1]);
      exit(1);
   }
   if(out==NULL){
      printf("Unable to open file %s for writing\n", argv[2]);
      exit(1);
   }

   int i = 0;

   // count lines of input file 
   while(fgets(line, MAX_LEN, in) != NULL){
      count++;
      token = strtok(line, " \n");
      while(token!=NULL){
         strcat(array[i], token);
         token = strtok(NULL, " \n");
         i++;
      }  
   }

   // Insertion Sort
   for(int i=0; i<count; i++){
      moveTo(L, 0);
      if(length(L)==0){
         append(L, 0);
      }else{
         while(1){
             if(strcmp(array[getElement(L)], array[i]) > 0){
                insertBefore(L, i);
                break;
             }else{
                moveNext(L);
                if(getIndex(L)==-1){
                   append(L, i);
                   break;
                }
             }
         }
      }
   }

   for(moveTo(L, 0); getIndex(L) != -1; moveNext(L)){
      fprintf(out, "%s\n", array[getElement(L)]);
   }


   // close files
   fclose(in);
   fclose(out);
   freeList(&L);
   return(0);
}

File: CMPS101/asg1/Lex.java
/**
   Name: Sean McGrath
   User ID: ssmcgrat
   Assignment: CMPS101 Programming Assignment 1
*/

import java.lang.*;
import java.io.*;
import java.util.Scanner;

class Lex{
   public static void main(String[] args) throws IOException{
      if( args.length != 2){
         System.err.println("Two arguments must be passed to Lex.java");
         return;
      }
    
      PrintWriter out = null; 
      int lineCount = 0;

      // Count number of lines in file
      try{ 
      lineCount = countLines(args[0]);
      }catch(FileNotFoundException e){
         System.err.println(e);
      }

      // Creates a string array from input file, each element in array is a line from file     
      String[] array = new String[lineCount];
      array = readFile(args[0], lineCount);
      

      // This is our second array for testing purposes
      String[] a = new String[lineCount];

      // Creates our integer list
      List L = InsertionSort(array);
      out = new PrintWriter(new FileWriter(args[1]));

      // Writing lines to output file
      for(int i=0; i<lineCount; i++){
         L.moveTo(i);
         int d = L.getElement();
         String line = array[d];
         out.println(line);
      }
      out.close();
   }

   // Method that sorts a String array and places the sorted array indeces into a List
   // returns the List
   //
   public static List InsertionSort(String[] a){
      List L = new List();
      L.append(0);
      L.moveTo(0);

      int i;
      int j;

      for(i=1; i<a.length; i++){
         L.moveTo(0);
         String temp = a[i];
         
         int elem  = L.getElement();
         j = i; 
         while( j>0 && temp.compareTo(a[elem])>0){
            L.moveNext();
            if(L.getIndex()!=-1){
               elem = L.getElement();
            }
            j--;
         }

         if(L.getIndex()==-1){
            L.append(i);
         } else{
            L.insertBefore(i);
            
         }

         
     }
      return L;
   }







   // Method that counts the number of lines in a file and returns that number.
   public static int countLines(String inFile) throws FileNotFoundException{
      int lineCount = 0;
      Scanner input = new Scanner(new FileReader(inFile));
      while(input.hasNextLine()){
         lineCount++;
         input.nextLine();
      }
      return lineCount;
   }

   // Method that reads lines from a file into an array of length "length" and returns array
   public static String[] readFile(String inFile, int length){
      String[] array = new String[length];
      try{
         FileInputStream fis = new FileInputStream(inFile);
         BufferedReader reader = new BufferedReader(new InputStreamReader(fis));

         String line = reader.readLine();
         for(int i=0; line!=null; i++){
            array[i] = line;
            line = reader.readLine();
         }
      }catch (FileNotFoundException e) {System.err.println(e);}
      catch (IOException ex){System.err.println(ex);}

      return array;      
   }

}
